### DISCUSSION TITLE
macOS touchpad pinch zoom in Avalonia 0.10

### DISCUSSION THREAD

**Discussion by arklumpus** [AvaloniaUI/Avalonia#13058] [Q&A] [answered]
Hi, is there a way to detect pinch zoom events from the touchpad in macOS in Avalonia 0.10? I cannot find any events that are fired when doing this.

This came up because I recently ran a demo of my software, and there was a user who would automatically use a pinch zoom motion even after I explained that it would not work and they needed to swipe up and down with two fingers...

**Comment by thevortexcloud** [comment]
I believe the pinch gesture was added in 11.0. 

https://docs.avaloniaui.net/docs/next/reference/gestures/pinchgesturerecognizer

If you have to do it in .10 you will probably need to implement it yourself. But it may be there.

**Reply by timunie** [reply]
touchpad != touchscreen. Iirc, we don't have a chance to know if the user has used the mouse device or trackpad. So no guestures available afaik.

**Reply by maxkatz6** [reply]
Yeah, macOS touchpad isn't technically a touch input, but a mouse. And has special events in both AppKit and Avalonia. (see my answer)

**Comment by maxkatz6** [comment, marked as answer]
There are a couple of special events that so far work only on macOS with touchpad:
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Input/Gestures.cs#L56-L66

But these were added only in 11.0-preview1 I believe, not in 0.10

**Reply by arklumpus** [reply]
I see, thank you!

I assume there is no reasonably easy way for me to backport this to 0.10 without touching the native assets, so I will just have to mark it as a known issue that should be fixed once I finally upgrade to 11.
